### PR TITLE
Rework HtmlTemplate from nested array to tree/OOP structure

### PR DESCRIPTION
**Pull request by mvorisek**
values are newly linked HtmlTemplate\Value objects and instanced only once per tag name

IMPORTANT:
- tree is dependent structure of parent template, ie. it can be modified with set/append/del operation on the parent template
- if another full template (instead of tree) is added, it is never affected with the owning template (except it can be deleted by it)
- new `Template::toLoadableString()` method is added
- data are checked if UTF-8 valid

BC:
- for user, all works like before
- except silently allowed bad templates with tag defined multiple times with different content - see changes (probably mistakes)
- `Template::get()` method is removed, the most closest equivalent is `Template::getTagTree()`, but the result type/structure is different
- `Template::eachTag()` method is removed as no more than one value per tag can exists - use `Template::getTagTree()` and walk/update tag tree as needed

PERFORMANCE:
- about the same in real world (parse/cache clone is about 4x slower, setters/getters are about 2x faster, render is about 2x slower)

`loadFromString`:
image

`renderToHtml`
image

### CODE REVIEW

**Review by DarkSide666 — COMMENTED** [review]
If it performs the same or better than current implementation, then I like it :)

File: demos/collection/jssortable.php [hunk 1]
```diff
@@ -13,7 +13,7 @@
     '<div class="ui header">Click and drag country to reorder</div>
     <div id="{$_id}" style="cursor: pointer">
         <ul>
-            {List}<li class="ui icon label" data-name="{$name}"><i class="{iso}ae{/} flag"></i> {name}andorra{/}</li>{/}
+            {List}<li class="ui icon label" data-name="{$name}"><i class="{iso}ae{/} flag"></i> {$name}</li>{/}
```
Comment (mvorisek) [inline demos/collection/jssortable.php:16]: old bug: tag defined twice with different content

File: demos/layout/templates/layout1.html [hunk 2]
```diff
@@ -123,12 +123,12 @@ <h4>Ungrouped transaction list</h4>{Report}
...
         <tfoot>
           <tr>
-            <th class="right aligned" colspan="6">Closing balance after {count}29{/} records:</th>
+            <th class="right aligned" colspan="6">Closing balance after {foot_count}29{/} records:</th>
             <th>€1,056.91</th>
             <th>€0.00</th>
           </tr>
           <tr>
-            <th class="right aligned" colspan="6">Total for {nominal}Sales:Limerick{/nominal}:</th>
+            <th class="right aligned" colspan="6">Total for {foot_nominal}Sales:Limerick{/}:</th>
```
Comment (mvorisek) [inline demos/layout/templates/layout1.html:131]: old bug: tag defined twice with different content

File: docs/template.rst [hunk 3]
```diff
@@ -717,8 +717,8 @@ under ``$template->template`::
 Property tags would contain::
 
     array (
-      'subject'=> array( &array ),
-      'subject#0'=> array( &array )
+      'subject#0'=> array( &array ),
```
Comment (mvorisek) [inline docs/template.rst:720]: old typo

File: tests/HtmlTemplateTest.php [hunk 4]
```diff
@@ -7,253 +7,168 @@
...
-    /**
-     * Test set() exception.
-     */
-    public function testSetException1()
+    public function testSetBadTypeException()
     {
         $t = new HtmlTemplate('{foo}hello{/} guys');
         $this->expectException(Exception::class);
-        $t->set('qwe', 'Hello'); // not existent tag
+        $t->set('foo', new \StdClass());
```
Comment (mvorisek) [inline tests/HtmlTemplateTest.php]:
```suggestion
        $t->set('foo', new \stdClass());
```

File: demos/layout/templates/layout1.html (hunk 2 above)
Comment (PhilippGrashoff) [reply demos/layout/templates/layout1.html:131]: If we are reworking this: Usually Regions are written with a capital letter `{Count}`, whereas Tags are written lowercase `{$something}`. Not mandatory, but good style.

Comment (mvorisek) [reply demos/layout/templates/layout1.html:131]: both are exactly the same :D

Comment (PhilippGrashoff) [reply demos/layout/templates/layout1.html:131]: yes, technically there is no difference, but in most ATK templates regions are capital, while pure tags (with a $) are written lowercase. As I said, not mandatory, but imo good style.

**Comment by mvorisek** [comment]
@georgehristov can you please review this PR?

**Review by georgehristov — APPROVED** [review]
LGTM

**Pull request merged**